Sources/SwiftProtobuf/FieldNameMap.swift

```diff
@@ -0,0 +1,105 @@
+// ProtobufRuntime/Sources/Protobuf/FieldNameMap.swift - Bidirectional field number/name mapping
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+
+/// An immutable bidirectional mapping between field names and numbers, used
+/// for various text-based serialization (JSON and text).
+public struct FieldNameMap: ExpressibleByDictionaryLiteral {
+
+  /// A "bundle" of names for a particular field. We use different "packed"
+  /// representations to minimize the amount of string data that we store in the
+  /// binary.
+  ///
+  /// TODO: Remove the Swift field name since it's only used for debugging
+  /// purposes and use the proto name instead at those usage sites.
+  public enum Names {
+
+    /// The proto name and the JSON name are the same string
+    case same(proto: String, swift: String)
+
+    /// The JSON and text names are different and not derivable from each other.
+    case unique(proto: String, json: String, swift: String)
+
+    // TODO: Add a case for JSON names that are computable from the proto name
+    // using the same algorithm implemented by protoc; for example,
+    // "foo_bar" -> "fooBar". This allows us to only store the string pointer
+    // in the payload once.
+
+    /// Returns the proto (and text format) name in the bundle.
+    public var protoName: String {
+      switch self {
+      case .same(proto: let name, swift: _): return name
+      case .unique(proto: let name, json: _, swift: _): return name
+      }
+    }
+
+    /// Returns the JSON name in the bundle.
+    public var jsonName: String {
+      switch self {
+      case .same(proto: let name, swift: _): return name
+      case .unique(proto: _, json: let name, swift: _): return name
+      }
+    }
+
+    /// Returns the Swift property name in the bundle.
+    public var swiftName: String {
+      switch self {
+      case .same(proto: _, swift: let name): return name
+      case .unique(proto: _, json: _, swift: let name): return name
+      }
+    }
+  }
+
+  /// The mapping from field numbers to name bundles.
+  private var numberToNameMap: [Int: Names] = [:]
+
+  /// The mapping from proto/text names to field numbers.
+  private var protoToNumberMap: [String: Int] = [:]
+
+  /// The mapping from JSON names to field numbers.
+  private var jsonToNumberMap: [String: Int] = [:]
+
+  /// Creates a new empty field name/number mapping.
+  public init() {}
+
+  /// Creates a new bidirectional field name/number mapping from the dictionary
+  /// literal.
+  public init(dictionaryLiteral elements: (Int, Names)...) {
```

Comment:
The compiler parses the `[ k: v, ... ]` literal and generates assembly code that creates a `ContiguousArray` of tuples. Since we can guarantee that the code generator generates them in sorted order, we *could* look at just preserving that array instead of building the number->name dictionary and then binary search it. But I think it's something we can revisit later.